Tests for TaskProvider #410

lemonboston opened the pull request:
Remove dynamic authority resolution in AuthorityUtil.

---

Submitting this now for first check. There are many TODOs left in code.

A few specific notes:
- opentaskpal classes are not reviewed yet
- the contentpal test support ended up being one simple `Matcher` class basically, but I used this `Check` type to implement it and make an adapter for `Matcher`. I haven't thought it through yet since things got much more simple with the `Assert` operations. One thing is that `Matcher` seems not to be completely straighforward for me with the 3 base classes, also I think what we do here is more an execution which can throw Exception than a simple comparison of an already calculated value, so that's why this interface may be better.
We could also forget the hamcrest assert as well, I suppose, to write something like:
```
execute(batch).with(client).andVerifyContent(assertOperations)
```
with our own types.
Implementation may depend a little bit on the eager queue stuff as well.
- not sure where the convenience `Assert*` composed operations should go, to contentpal module or contentpal-testing module
- I've left `AA_TO_*` package names to indicate which module the sibling packages are intended to end up in
- There is one bigger test case left in original form, I will update that, too, so the related private methods can be removed as well
- All test cases are based on the original ones from the old code
- There are possible ways to improve the test code usage, there are some issues to be solved as well (TODOs in code)
- `Assert` for `RowSnapshot` is not available yet, so that will result in changes as well

This is a single commit now, I will continue to commit on top of it.

dmfs commented on opentaskspal/src/main/java/org/dmfs/opentaskspal/operations/Task.java:
```diff
@@ -0,0 +1,57 @@
+import org.dmfs.tasks.contract.TaskContract;
+
+
+public final class Task implements InsertOperation<TaskContract.Tasks>
```
The name is not ideal. This is a decorator which puts new a task on a specific list. The name should express that. Also it's valid to change the task list a list belongs to so this operation is not limited to inserts.

lemonboston replied:
This comes back to the question of whether to keep `InsertOperation` and `Operation`, because if we keep them, than we need to make 2 classes here for the same thing since it cannot be used in `Table.insertOpertation()` for return vallue if it's an `Operation`.
For now, I will just leave it as an `InsertOperation` and rename it.

dmfs commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/testing/Check.java:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @author Gabor Keszthelyi
+ */
+public interface Check<T>
```
Do we really need this interface at all?

lemonboston replied:
Since we basically only have a single and simple `Matcher` now I suppose we don't, so I've removed it. I still think that it may be nice to have a custom flowable 'api' to test operations, but it's not important, we could get back to it later as well.

lemonboston commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/predicates/DelegatingPredicate.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * @author Gabor Keszthelyi
+ */
+public abstract class DelegatingPredicate implements Predicate
```
Yep, all class in the `org.dmfs.android.contentpal` package here are intended to end up there, so I start to create issue for them now. Done for this one.

lemonboston commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/predicates/IdEq.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public final class IdEq<PrimaryTable> extends DelegatingPredicate
+{
+    public IdEq(final String foreignIdColumnName, final RowSnapshot<PrimaryTable> primaryRow, final String primaryIdColumnName)
```
You're right, that name is much better and there is no need to specify the primary id column.
Submitted pull request in contentpal: https://github.com/dmfs/ContentPal/pull/47

lemonboston commented on opentaskspal/src/main/java/org/dmfs/opentaskspal/predicates/ListIdEq.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author Gabor Keszthelyi
+ */
+public final class ListIdEq extends DelegatingPredicate
```
Ok, renamed it to `TaskOnList`.

dmfs commented on opentasks-provider/build.gradle:
```diff
@@ -17,11 +17,26 @@ android {
+
+    // https://github.com/amitshekhariitbhu/Android-Debug-Database
+    // Usage: "adb forward tcp:8080 tcp:8080" then open "http://localhost:8080"
+    androidTestCompile 'com.amitshekhar.android:debug-db:1.0.1'
```
What's this for and does it have to be in master?

lemonboston replied:
This was very useful for me, it allows to check the content of the db from browser. Do you maybe use something else for that?
Anyway, I've removed it, no need for it to be committed.

dmfs commented on opentaskspal/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,9 @@
+        package="org.dmfs.opentaskspal">
+
+    <!--TODO Remove after https://github.com/dmfs/opentasks/issues/392-->
+    <uses-sdk tools:overrideLibrary="org.dmfs.android.contentpal"/>
```
please add a note to #392 about this.

lemonboston replied:
added: https://github.com/dmfs/opentasks/issues/392#issuecomment-329198715

dmfs commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/operations/AssertNoRow.java:
```diff
@@ -0,0 +1,33 @@
+{
+    public AssertNoRow(Table<T> table, RowData<T> rowData)
+    {
+        super(new Counted<T>(0, new BulkAssert<T>(table, rowData)));
```
Again, this is the same as `AssertEmptyTable`. It always fails if the given `Table` is not empty.

lemonboston replied:
This has been removed already.

dmfs commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/operations/AssertRow.java:
```diff
@@ -0,0 +1,43 @@
+    public AssertRow(Table<T> table, RowSnapshot<T> rowSnapshot, RowData<T> rowData)
+    {
+        super(new Counted<T>(1,
+                new BulkAssert<T>(table,
```
This should use `Assert` now.

lemonboston replied:
This was removed as well while you were reviewing.

dmfs commented:
As mentioned before I see a general problem with the "RelatedTo" type of Predicates. They won't work if the row has not been loaded from the database. Because the `VirtualRowSnapshot` doesn't contain the `_ID` it got after it had been committed. That's why this type of constraint needs to be added by an `Operation` decorator, because it has access to the `TransactionContext`.

dmfs commented:
The test project should use a different provider authority like `org.dmfs.opentasks.test` to allow testing even if OpenTasks is already installed on the test device.

lemonboston commented:
Test not uses different authority name (added a resource file under `androidTest` to override the string resource).

lemonboston commented on opentaskspal/src/main/java/org/dmfs/android/contentpal/operations/AssertNoRow.java:
```diff
@@ -0,0 +1,33 @@
+{
+    public AssertNoRow(Table<T> table, RowData<T> rowData)
+    {
+        super(new Counted<T>(0, new Assert<T>(table, rowData)));
```
(Addressed already)

lemonboston commented:
(Corrected typo in previous comment: "Test not uses.." -> "Test now uses..")

lemonboston commented:
OpenTaksPal classes are ready for review (i.e. classes in `opentaskpal` module's `org.dmfs.opentaskspal` package. The ones in `org.dmfs.android.contentpal` package are all supposed to be replaced with ones from ContentPal.)

lemonboston commented:
This is now rebased on top of https://github.com/dmfs/opentasks/pull/422, I've changed the title as well.
`PostOperationAssert*` classes are still used here, but could be replaced with `BulkAssert(..ReferredTo())` from contentpal PR if needed as mentioned on Slack.

dmfs commented on opentasks-provider/src/androidTest/java/org/dmfs/provider/tasks/TaskProviderTest.java:
```diff
@@ -0,0 +1,404 @@
+        RowSnapshot<Tasks> task = new VirtualRowSnapshot<>(taskTable);
+        OperationsQueue queue = new BasicOperationsQueue(mClient);
+
+        assertThat(new MultiBatch(
```
There is no need to assert this again if it already has been verified by a previous test. Just insert the rows.

lemonboston replied:
Removed that unnecessary assert.

dmfs commented:
@lemonboston please rebase on master

lemonboston commented:
Rebased on master and updated tests with latest contentpal classes, finalized the test code, added javadocs for all test methods.

To sum up, this pull request adds Instrumentation tests for TaskProvider, based on test cases from earlier times of the project. (ticket for further tests: https://github.com/dmfs/opentasks/issues/420)
It also removes the dynamic authority resolution.

There are 2 TODOs left in the code, let me know if either is worth having a ticket:
- Test Orchestration is a coming feature which will allow each test method to run in its own instrumentation, hence removing the need for some code in setUp() and tearDown()
- Authority is now simply cached under the same class, AuthorityUtil, design or name could be improved (results in changes in many files)

dmfs commented:
Yes, please create tickets.

lemonboston commented:
Created the tickets for the TODOs: https://github.com/dmfs/opentasks/issues/437, https://github.com/dmfs/opentasks/issues/438
